07_distributed_systems/01_intro/client/client.go

```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"kv_store/protocol"
+	"net/http"
+	"os"
+	"regexp"
+	"time"
+)
+
+const (
+	SERVER = "127.0.0.1"
+	PORT   = "8000"
+	URL    = "http://" + SERVER + ":" + PORT
+)
+
+func panicOnError(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
+
+// abozhenko for oz: Is it ok to use http at all?
+
+func get(key string) (response string, err error) {
+	//TODO: re-use tcp connection for entire repl session?
+	getRequest := protocol.GetRequest{
+		Key: []byte(key),
+	}
+	reqBytes := getRequest.Encode()
+
+	//abozhenko for oz: Do you think it is ok to use get request body?
+	// I saw some discussion here
+	// https://stackoverflow.com/questions/978061/http-get-with-request-body
+	// but I haven't form my own opinion
+	req, err := http.NewRequest(http.MethodGet, URL+"/get",
+		bytes.NewReader(reqBytes))
+	panicOnError(err)
+	req.Header.Set("Content-Type", "application/octet-stream")
+
+	start := time.Now()
+	resp, err := http.DefaultClient.Do(req)
+	elapsed := time.Since(start)
+	panicOnError(err)
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	return string(body) + " <" + fmt.Sprintf("%4.2f", float64(elapsed)/1_000_000.0) + "ms>", err
+}
+
+func set(key, value string) (response string, err error) {
+	setRequest := protocol.SetRequest{
+		Key:   []byte(key),
+		Value: []byte(value),
+	}
+	reqBytes := setRequest.Encode()
+	// bytes.NewReader copies the []byte, right?
+	// No, it does not! https://stackoverflow.com/a/39993797/1572363
+	// You should've known.
+
+	// abozhenko for oz: Is it ok to simply use different endpoints for
```

Comment:
@alexbozhenko it's not the most elegant, and against the REST philosophy of different HTTP verbs determining behavior at the same endpoint, but if you're not necessarily following REST it's fine